Computed value CSS gradients need to serialize colors in their computed form

weinig opened the pull request:
#### 6c434b5464a00be355e52531b1d20d2637e2b295
<pre>
Computed value CSS gradients need to serialize colors in their computed form
<a href="https://bugs.webkit.org/show_bug.cgi?id=246927">https://bugs.webkit.org/show_bug.cgi?id=246927</a>

Reviewed by Darin Adler.

Converts StyleGradientImage to store color stop colors as StyleColors rather
than keeping the primitive values. This is the intended model of the Style
classes (storing Style types rather than CSS types) and is now possible that
the spec requires that the computed value not maintain the specified values
(e.g. we use `rgb(255, 0, 0)` and not `red` if `red` was specified).

* LayoutTests/fast/gradients/css3-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/css3-gradient-parsing.html:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing.html:
Update tests to check both specified and computed forms.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-image-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/animations/border-image-source-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-image-source-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/parsing/gradient-interpolation-method-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/animations/list-style-image-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/mask-image-interpolation-expected.txt:
Update results for new serialization behavior,.

* LayoutTests/imported/w3c/web-platform-tests/css/css-images/parsing/gradient-interpolation-method-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-substitution-background-properties.html:
Import new version of the tests that checks the new computed form.

* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::computeStyleGradientImageStopColor):
(WebCore::CSSGradientValue::computeStops const):
Convert to StyleImage.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::resolveColorWithoutCurrentColor const): Deleted.
* Source/WebCore/css/StyleColor.h:
Remove unused (and dangerous if used wrong) function.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
Update for simplified StyleGradientImage::Stop.

* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::operator==):
(WebCore::stopsAreCacheable):
(WebCore::resolveColorStopColor):
(WebCore::StyleGradientImage::StyleGradientImage):
(WebCore::computedStyleValueForColorStopColor):
(WebCore::StyleGradientImage::computedStyleValue const):
(WebCore::StyleGradientImage::image const):
(WebCore::StyleGradientImage::knownToBeOpaque const):
(WebCore::StyleGradientImage::computeStops const):
(WebCore::StyleGradientImage::hasColorDerivedFromElement const): Deleted.
(WebCore::StyleGradientImage::isCacheable const): Deleted.
* Source/WebCore/rendering/style/StyleGradientImage.h:
Rathen than storing a CSSPrimitiveValue for the stop color, we can use
a StyleColor and remove the extra Color member entirely. Now that we
don&apos;t have the CSSPrimitiveValue, we need to compute the CSS value for
the color stops when creating a computed value. This happens to give
us the correct new behavior of having them fully resolved.

* Source/WebCore/style/StyleBuilderState.h:
Remove function definition with no body.

Canonical link: <a href="https://commits.webkit.org/256073@main">https://commits.webkit.org/256073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d93273038315f082def94ab8b496e677f2d06d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3713 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27439 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104187 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164459 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98535 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3768 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31909 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86864 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100207 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80926 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29742 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/72637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38310 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36173 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4195 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40073 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42028 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/38528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/fast/gradients/css3-gradient-parsing.html:
```diff
@@ -9,67 +9,111 @@
 }
 
-function testSame(dir, prop, altProp, value)
+function testGradientComputedValue(css, queryProp)
```
Starting to wish these tests were in Web Platform Tests. I find it easier to make a good test in our own tests, but the power to drive interoperability when the tests are in WPT is so worthwhile. Maybe we can get someone to convert these?

weinig replied:
I am just putting it on my list of things to upstream. There is a whole parsing section in the css-images WPT folder that I already have contributed a bunch of test to.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-image-computed.sub-expected.txt:
```diff
@@ -2,35 +2,35 @@
-FAIL Property background-image value 'radial-gradient(circle calc(-0.5em + 10px) at calc(-1em + 10px) calc(-2em + 10px), red, blue)' assert_equals: expected "radial-gradient(0px at -30px -70px, rgb(255, 0, 0), rgb(0, 0, 255))" but got "radial-gradient(calc(-0.5em + 10px) at calc(-1em + 10px) calc(-2em + 10px), red, blue)"
-FAIL Property background-image value 'radial-gradient(ellipse calc(-0.5em + 10px) calc(0.5em + 10px) at 20px 30px, red, blue)' assert_equals: expected "radial-gradient(0px 30px at 20px 30px, rgb(255, 0, 0), rgb(0, 0, 255))" but got "radial-gradient(calc(-0.5em + 10px) calc(0.5em + 10px) at 20px 30px, red, blue)"
-FAIL Property background-image value 'radial-gradient(ellipse calc(0.5em + 10px) calc(-0.5em + 10px) at 20px 30px, red, blue)' assert_equals: expected "radial-gradient(30px 0px at 20px 30px, rgb(255, 0, 0), rgb(0, 0, 255))" but got "radial-gradient(calc(0.5em + 10px) calc(-0.5em + 10px) at 20px 30px, red, blue)"
+FAIL Property background-image value 'radial-gradient(circle calc(-0.5em + 10px) at calc(-1em + 10px) calc(-2em + 10px), red, blue)' assert_equals: expected "radial-gradient(0px at -30px -70px, rgb(255, 0, 0), rgb(0, 0, 255))" but got "radial-gradient(calc(-0.5em + 10px) at calc(-1em + 10px) calc(-2em + 10px), rgb(255, 0, 0), rgb(0, 0, 255))"
```
Surprised cases like this didn’t progress.

weinig replied:
A follow on when I do the same thing I am doing here for color for the other values in gradients will fix the rest of this.

darinadler commented on Source/WebCore/css/CSSGradientValue.cpp:
```diff
@@ -36,20 +36,18 @@
+}
 
-    return stops;
+Vector<StyleGradientImageStop> CSSGradientValue::computeStops(Style::BuilderState& state) const
```
So sad to have to utter `StyleGradientImageStop` three times here. Glad we don’t have to do it four times, though!

weinig replied:
We don't have to if we don't want to!

```c++
static inline std::optional<StyleColor> computeStyleColor(const RefPtr<CSSPrimitiveValue>& color, Style::BuilderState& state)
{
    if (!color)
        return std::nullopt;
    return state.colorFromPrimitiveValue(*color);
}

decltype(auto) CSSGradientValue::computeStops(Style::BuilderState& state) const
{
    return m_stops.map([&] (auto& stop) -> StyleGradientImageStop {
        return { computeStyleColor(stop.color, state), stop.position };
    });
}
```

This works just as well :).

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.h:
```diff
@@ -34,9 +34,8 @@
 
 struct StyleGradientImageStop {
-    RefPtr<CSSPrimitiveValue> color;
+    std::optional<StyleColor> color;
```
Both `Color` and `StyleColor` have "invalid" values, so I was hoping we’d almost never use `std::optional<StyleColor>` just as we wouldn’t use `std::optional<RefPtr<Node>>` since `nullptr` is there and handy. On the other hand, doing it explicitly like this is so nice. Maybe some day we could change `StyleColor` and `Color` to not have invalid states when used directly, and only have them when combined with a template analogous to `std::optional`; lets us be explicit about whether there is a null, but without extra storage overhead.

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.h:
```diff
@@ -103,15 +102,12 @@ class StyleGradientImage final : public StyleGeneratedImage {
-    Vector<Stop> m_stops;
-    mutable std::optional<bool> m_hasColorDerivedFromElement;
+    Vector<StyleGradientImageStop> m_stops;
+    bool m_knownCacheableBarringFilter;
```
Would be nice to initialize this here even if that’s never relied on in practice. I would suggest we do it with all scalars unless we are trying to optimize something.

weinig replied:
Seems reasonable. I kinda wish we (and maybe this exists) we could just enforce that all fields must be initialized at the end of construction (with some ability to disable when needed), but maybe c++ initialization patterns are too complicated / cross translation unity to allow for that kind of analysis (Swift has made me soft).

anttijk commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -186,12 +215,10 @@ RefPtr<Image> StyleGradientImage::image(const RenderElement* renderer, const Flo
 
 bool StyleGradientImage::knownToBeOpaque(const RenderElement& renderer) const
 {
-    bool hasColorFilter = renderer.style().hasAppleColorFilter();
+    auto& style = renderer.style();
+    bool hasColorFilter = style.hasAppleColorFilter();
```
I think the issue with the failing ::first-line test is that these renderer taking functions (most importantly image()) should be using `firstLineStyle()` instead of `style()` when appropriate. Whether we are on the first line could be passed as an additional boolean argument, or style itself could be passed (though at least some of these use the renderer for other things too so it would also be needed).  

The only place (?) where the first-line bit needs to be passed is in `BackgroundPainter::paintFillLayer()` where you would test with something like `bool isFirstLine = box && box->lineBox()->isFirst();`

weinig replied:
Thanks Antti! To keep things a bit sane, I am going to do that in a follow up, and make a small revert here to do the currentColor lookup in the old place. Filed https://bugs.webkit.org/show_bug.cgi?id=247127 for this.

Commit:
Computed value CSS gradients need to serialize colors in their computed form
https://bugs.webkit.org/show_bug.cgi?id=246927

Reviewed by Darin Adler.

Converts StyleGradientImage to store color stop colors as StyleColors rather
than keeping the primitive values. This is the intended model of the Style
classes (storing Style types rather than CSS types) and is now possible that
the spec requires that the computed value not maintain the specified values
(e.g. we use `rgb(255, 0, 0)` and not `red` if `red` was specified).

* LayoutTests/fast/gradients/css3-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/css3-gradient-parsing.html:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing-expected.txt:
* LayoutTests/fast/gradients/unprefixed-gradient-parsing.html:
Update tests to check both specified and computed forms.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-image-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/animations/border-image-source-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-image-source-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/parsing/gradient-interpolation-method-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/animations/list-style-image-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-image-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/mask-image-interpolation-expected.txt:
Update results for new serialization behavior,.

* LayoutTests/imported/w3c/web-platform-tests/css/css-images/parsing/gradient-interpolation-method-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-substitution-background-properties.html:
Import new version of the tests that checks the new computed form.

* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::computeStyleGradientImageStopColor):
(WebCore::CSSGradientValue::computeStops const):
Convert to StyleImage.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::resolveColorWithoutCurrentColor const): Deleted.
* Source/WebCore/css/StyleColor.h:
Remove unused (and dangerous if used wrong) function.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
Update for simplified StyleGradientImage::Stop.

* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::operator==):
(WebCore::stopsAreCacheable):
(WebCore::resolveColorStopColor):
(WebCore::StyleGradientImage::StyleGradientImage):
(WebCore::computedStyleValueForColorStopColor):
(WebCore::StyleGradientImage::computedStyleValue const):
(WebCore::StyleGradientImage::image const):
(WebCore::StyleGradientImage::knownToBeOpaque const):
(WebCore::StyleGradientImage::computeStops const):
(WebCore::StyleGradientImage::hasColorDerivedFromElement const): Deleted.
(WebCore::StyleGradientImage::isCacheable const): Deleted.
* Source/WebCore/rendering/style/StyleGradientImage.h:
Rathen than storing a CSSPrimitiveValue for the stop color, we can use
a StyleColor and remove the extra Color member entirely. Now that we
don't have the CSSPrimitiveValue, we need to compute the CSS value for
the color stops when creating a computed value. This happens to give
us the correct new behavior of having them fully resolved.

* Source/WebCore/style/StyleBuilderState.h:
Remove function definition with no body.

Canonical link: https://commits.webkit.org/256073@main
